### PR TITLE
Qe

### PR DESCRIPTION
**Pull request by mhrmm**
@sbhaktha Here's the PR I talked about. It has yours already merged in. It also refactors some of the reranking CLIs, in addition to merging in Michael's Stanford POS tagger change. Finally, it speeds up the file reading for parse trees.

### CODE REVIEW

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala [hunk 1]
```diff
@@ -65,12 +75,86 @@ case object FactorieSentenceTagger extends SentenceTransform {
...
+          'autoCpos -> Set(autoCpos)
+        ))
+    }))
+  }
+}
+
+/** The MultiSentenceTagger tags an input sentence with automatic part-of-speech tags
+  * from the Stanford tagger.
+  */
+case object MultiSentenceTagger extends SentenceTransform {
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: `MultiSentenceTagger` sounds like it tags multiple sentences. Rename? To `SentenceMultiTagger` or something?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/GoldParseTrainingVectorSource.scala [hunk 2]
```diff
@@ -68,6 +68,9 @@ case class GoldParseTrainingVectorSource(
     for {
       goldParse <- goldParses.parseIterator
       vector <- generateVectors(goldParse)
-    } yield vector
+    } yield {
+      //println(vector)
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/GoldParseTrainingVectorSource.scala]: Looks like diagnostic print is commented out. Revert back to `yield vector` the way it was?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala [hunk 3]
```diff
@@ -65,12 +75,86 @@ case object FactorieSentenceTagger extends SentenceTransform {
...
+  def transform(sentence: Sentence): Sentence = {
+    val words: IndexedSeq[String] = sentence.tokens.tail map { tok => tok.word.name }
+    val nlpStackTokens: IndexedSeq[NLPStackToken] =
+      Tokenizer.computeOffsets(words, words.mkString).toIndexedSeq
+    val stanfordTaggedTokens: IndexedSeq[PostaggedToken] =
+      stanfordTagger.postagTokenized(nlpStackTokens).toIndexedSeq
+    val factorieTaggedTokens: IndexedSeq[PostaggedToken] =
+      defaultPostagger.postagTokenized(nlpStackTokens).toIndexedSeq
+    Sentence(NexusToken +: (stanfordTaggedTokens.zip(factorieTaggedTokens).zip(sentence.tokens.tail) map {
+      case ((stanTagged, factTagged), untagged) =>
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: Can we call these `stanfordTagged` and `factorieTagged` ? `factTagged` is a little confusing.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/PolytreeParse.scala [hunk 4]
```diff
@@ -313,7 +314,7 @@ object PolytreeParse {
     makePoly: Boolean): Iterator[PolytreeParse] = {
 
     val rawIter =
-      fromConllHelper(Source.fromFile(filename).getLines, useGoldPosTags = useGoldPosTags).iterator
+      fromConllHelper(Source.fromFile(filename).getLines, useGoldPosTags = useGoldPosTags) //.iterator
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/PolytreeParse.scala]: Remove commented `.iterator`

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/PolytreeParse.scala [hunk 5]
```diff
@@ -347,30 +348,39 @@ object PolytreeParse {
...
-      if (rest.hasNext) {
-        // check for more in case last line in file is newline
-        fromConllHelper(rest, useGoldPosTags)
-      } else {
-        Stream.empty
+  ): Iterator[PolytreeParse] = {
+
+    // Groups the lines of the file by parse (parses are identified via a blank line
+    // in the file).
+    def groupby[T](iter: Iterator[T])(startsGroup: T => Boolean): Iterator[Iterator[T]] =
```
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/PolytreeParse.scala]: What's issue from the previous code is the `groupBy` implementation trying to fix?

**Comment by sbhaktha** [comment]
Looks good. A few suggestions wrt names used. Also couldn't make sense of the `groupBy` change -- had a question. Otherwise cool.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/PolytreeParse.scala (hunk 5 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/PolytreeParse.scala]: Speed. It was really slow for long files, for some reason. My theory is that my streaming logic was somehow done incorrectly (though I don't honestly know why the other implementation was so slow).

**Comment by mhrmm** [comment]
Revised! Merging.

**Pull request merged**